tests/scene/test_image_texture.h

```diff
@@ -0,0 +1,121 @@
+	CHECK(image_texture->has_alpha() == true);
+	CHECK(image_texture->get_rid().is_valid() == true);
+}
+
+TEST_CASE("[SceneTree][ImageTexture] set_image") {
+	Ref<ImageTexture> image_texture = memnew(ImageTexture);
+	Ref<Image> image = memnew(Image(8, 4, false, Image::FORMAT_RGB8));
+	image_texture->set_image(image);
+	CHECK(image_texture->get_width() == 8);
+	CHECK(image_texture->get_height() == 4);
+	CHECK(image_texture->get_format() == Image::FORMAT_RGB8);
+	CHECK(image_texture->has_alpha() == false);
+	CHECK(image_texture->get_width() == image_texture->get_image()->get_width());
+	CHECK(image_texture->get_height() == image_texture->get_image()->get_height());
+	CHECK(image_texture->get_format() == image_texture->get_image()->get_format());
+}
+
+TEST_CASE("[SceneTree][ImageTexture] set_size_override") {
+	Ref<Image> image = memnew(Image(16, 8, false, Image::FORMAT_RGB8));
+	Ref<ImageTexture> image_texture = ImageTexture::create_from_image(image);
+	CHECK(image_texture->get_width() == 16);
+	CHECK(image_texture->get_height() == 8);
+	image_texture->set_size_override(Size2i(32, 16));
+	CHECK(image_texture->get_width() == 32);
+	CHECK(image_texture->get_height() == 16);
+}
+
+TEST_CASE("[SceneTree][ImageTexture] is_pixel_opaque") {
+	Ref<Image> image = memnew(Image(8, 8, false, Image::FORMAT_RGBA8));
+	image->set_pixel(0, 0, Color(0.0, 0.0, 0.0, 0.0)); // not opaque
+	image->set_pixel(0, 1, Color(0.0, 0.0, 0.0, 0.1)); // not opaque
+	image->set_pixel(0, 2, Color(0.0, 0.0, 0.0, 0.5)); // opaque
+	image->set_pixel(0, 3, Color(0.0, 0.0, 0.0, 0.9)); // opaque
+	image->set_pixel(0, 4, Color(0.0, 0.0, 0.0, 1.0)); // opaque
+
+	Ref<ImageTexture> image_texture = ImageTexture::create_from_image(image);
+	CHECK(image_texture->is_pixel_opaque(0, 0) == false);
+	CHECK(image_texture->is_pixel_opaque(0, 1) == false);
+	CHECK(image_texture->is_pixel_opaque(0, 2) == true);
+	CHECK(image_texture->is_pixel_opaque(0, 3) == true);
+	CHECK(image_texture->is_pixel_opaque(0, 4) == true);
+}
+
+TEST_CASE("[SceneTree][ImageTexture] set_path") {
+	Ref<ImageTexture> image_texture = memnew(ImageTexture);
+	image_texture->set_path(TestUtils::get_data_path("images/icon.png"), true);
+	CHECK(image_texture->get_path() == TestUtils::get_data_path("images/icon.png"));
+}
+
+TEST_CASE("[SceneTree][ImageTexture] draw") {
+	Ref<Image> image = Image::load_from_file(TestUtils::get_data_path("images/icon.png"));
+	Ref<ImageTexture> image_texture = ImageTexture::create_from_image(image);
+	RID canvas = RenderingServer::get_singleton()->canvas_create();
+	RID canvas_item = RenderingServer::get_singleton()->canvas_item_create();
+	RenderingServer::get_singleton()->canvas_item_set_parent(canvas_item, canvas);
+	image_texture->draw(canvas_item, Point2(0.0, 0.0));
+	// TODO: figure out how to check to see if it was drawn. can we sample a pixel on the canvas?
+	// I suppose not segfaulting/crashing is a positive result that might make this worth
+	// keeping.
+}
```

Comment:
The tests typically run as headless, but they should also work if ran with an actual rendering server and display server.

So it's worth making sure that it does something useful in non-headless mode, and if it can't work in headless, it can be made optional with some `if (DisplayServer::get_singleton()->get_name() == "headless")` checks.

Now I didn't check, but does headless drawing to a texture in memory actually work? If it does, that's also worth testing.

Reply:
I'm not sure how to execute tests headless vs non-headless. When you add the "--test" flag, the editor binary doesn't actually start up, it just executes the tests and exits like the console binary. Looking at the startup code, it looks like if the "--test" argument is present, the main enters "TEST_MAIN_PARAM_OVERRIDE" and returns/exits from that before opening a windows context or running the actual proper engine main.

Thins I've tried:
 * `DisplayServer::get_singleton()->get_name()` returns "mock"
 * `DisplayServer::get_singleton()->window_can_draw()` returns false
 * Adding proper viewport initialization code (see below), errors out on "RenderDummy" calls to access a null pointer viewport texture.
  * Accessing "RSG::viewport->get_total_objects_drawn()" and similar metrics functions, they all always return 0.

Unless anyone has any other ideas, I'm on board to just delete the draw test case.
   
```
TEST_CASE("[SceneTree][ImageTexture] draw") {
	print_line(DisplayServer::get_singleton()->get_name()); // prints "mock"

	RID scenario = RS::get_singleton()->scenario_create();
	RID viewport = RS::get_singleton()->viewport_create();
	RS::get_singleton()->viewport_set_update_mode(viewport, RS::VIEWPORT_UPDATE_DISABLED);
	RS::get_singleton()->viewport_set_scenario(viewport, scenario);
	RS::get_singleton()->viewport_set_size(viewport, 128, 128);
	RS::get_singleton()->viewport_set_transparent_background(viewport, true);
	RS::get_singleton()->viewport_set_active(viewport, true);

	Ref<Image> image = Image::load_from_file(TestUtils::get_data_path("images/icon.png"));
	Ref<ImageTexture> image_texture = ImageTexture::create_from_image(image);
	RID canvas = RenderingServer::get_singleton()->canvas_create();
	RID canvas_item = RenderingServer::get_singleton()->canvas_item_create();
	RenderingServer::get_singleton()->canvas_item_set_parent(canvas_item, canvas);
	image_texture->draw(canvas_item, Point2(0.0, 0.0));
	
	print_line("get_total_objects_drawn: ", RSG::viewport->get_total_objects_drawn()); // prints 0
	print_line("get_total_primitives_drawn: ", RSG::viewport->get_total_primitives_drawn()); // prints 0
	print_line("get_total_draw_calls_used: ", RSG::viewport->get_total_draw_calls_used()); // prints 0
	
	RS::get_singleton()->viewport_attach_canvas(viewport, canvas);
	RID viewport_texture = RS::get_singleton()->viewport_get_texture(viewport);
	Ref<Image> img = RS::get_singleton()->texture_2d_get(viewport_texture); // ERROR: Parameter "t" is null.
// at: RendererDummy::TextureStorage::texture_2d_get (C:\repos\godot\servers\rendering\dummy\storage\texture_storage.h:106)
 	if (img.is_valid() && !img->is_empty()) { 
		// never reached
	} else {
		print_line("viewport texture is null/empty"); // executed every time
	}
}
```

Reply:
> Unless anyone has any other ideas, I'm on board to just delete the draw test case.

Yeah, makes sense to just remove it. If desired (and decided what's a proper way to test it) it could be always added separately in another PR.
(Sorry for the late response; I've seen your comment before but I think I've missed this specific sentence. :thinking:)

Reply:
All good, my comment was a bit of a wall of text. I went ahead and deleted the test case. I'll re-request a review.

Reply:
> I went ahead and deleted the test case. I'll re-request a review.

I've already approved the PR after the removal. So that was a pointless re-request. :upside_down_face: